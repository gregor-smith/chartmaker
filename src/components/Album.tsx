import React, { FC } from 'react'
import { css, cx } from 'emotion'
import { Album as AlbumDetails, formatAlbumTitle } from '../state'


type Props = {
    details: AlbumDetails | null
    sizeRem: number
}


const baseStyle = css({
    backgroundColor: 'white',
    margin: '0.15rem'
})


export const Album: FC<Props> = ({ details, sizeRem }) => {
    const style = cx(
        baseStyle,
        css({
            width: `${sizeRem}rem`,
            height: `${sizeRem}rem`
        })
    )
    const image = details === null
        ? null
        : <img src={details.image.smallURL} alt={formatAlbumTitle(details)}/>
    return (
        <div className={style}>
            {image}
        </div>
    )
}
