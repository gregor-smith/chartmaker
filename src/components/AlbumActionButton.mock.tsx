const mock: typeof import('@/components/AlbumActionButton') = {
    AlbumActionButton: props =>
        <div {...props} className='mock-album-action-button'/>
}


export const AlbumActionButton = mock.AlbumActionButton
