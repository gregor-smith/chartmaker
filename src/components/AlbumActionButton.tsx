import { h, FunctionComponent } from 'preact'
import { css } from 'emotion'


type Props = {
    onClick: () => void
    title: string
}


const style = css({
    background: 'none',
    border: 'none'
})


export const AlbumActionButton: FunctionComponent<Props> = props =>
    <button {...props} class={style}/>
