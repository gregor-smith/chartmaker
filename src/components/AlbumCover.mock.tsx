import { cx } from 'emotion'


const mock: typeof import('@/components/AlbumCover') = {
    AlbumCover: ({
        album,
        size,
        overlayClass,
        children,
        highlighted,
        ...props
    }) => {
        const json = JSON.stringify(album)
        const className = cx('mock-album-cover', overlayClass)
        return (
            <div {...props} className={className}>
                {`Album: ${json}`}
                {`Size: ${size}`}
                {`Highlighted: ${highlighted}`}
                {children}
            </div>
        )
    }
}


export const AlbumCover = mock.AlbumCover
