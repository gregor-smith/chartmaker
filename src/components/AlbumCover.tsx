import React, { FC, DragEvent } from 'react'
import { css, cx } from 'emotion'

import { TEXT_COLOUR, ALBUM_PADDING_SIZE, highlightBackgroundStyle } from '@/style'
import { Album } from '@/types'


export type AlbumCoverProps = {
    album: Album
    size: string
    onDragStart: (event: DragEvent<HTMLDivElement>) => void
    onDragOver?: (event: DragEvent<HTMLDivElement>) => void
    onDragEnter?: (event: DragEvent<HTMLDivElement>) => void
    onDrop?: (event: DragEvent<HTMLDivElement>) => void
    onMouseEnter?: () => void
    overlayClass?: string
    highlighted?: boolean
}


const baseContainerStyle = css({
    position: 'relative',
    background: TEXT_COLOUR,
    margin: ALBUM_PADDING_SIZE
})

const imageStyle = css({
    position: 'absolute',
    top: 0,
    left: 0,
    width: '100%',
    maxWidth: '100%',
    height: '100%',
    objectFit: 'contain'
})

// the overlay is given an additional non-emotion class so it can be used as a
// selector in the container's style
const overlaySelector = 'overlay'

const baseOverlayStyle = cx(
    overlaySelector,
    imageStyle,
    css({ zIndex: 10 })
)


export const AlbumCover: FC<AlbumCoverProps> = ({
    album,
    size,
    onDragStart,
    onDragEnter,
    onDragOver,
    onDrop,
    onMouseEnter,
    children,
    overlayClass,
    highlighted
}) => {
    const overlayStyle = cx(baseOverlayStyle, overlayClass)
    const containerStyle = cx(
        baseContainerStyle,
        css({
            width: size,
            height: size,
            [`:not(:hover) .${overlaySelector}`]: {
                display: 'none'
            }
        }),
        highlighted === false
            ? highlightBackgroundStyle
            : undefined
    )

    let image: JSX.Element | undefined
    if (!album.placeholder) {
        image = (
            // lazy loading causes problems with html2canvas so keep it default
            <img className={imageStyle}
                src={album.url}
                alt={album.name}
                crossOrigin='anonymous'/>
        )
    }

    return (
        <div className={containerStyle}
                draggable={!album.placeholder}
                onDragStart={onDragStart}
                onDragEnter={onDragEnter}
                onDragOver={onDragOver}
                onDrop={onDrop}
                onMouseEnter={onMouseEnter}
                title={album.placeholder ? undefined : album.name}>
            {image}
            <div className={overlayStyle}>
                {children}
            </div>
        </div>
    )
}
