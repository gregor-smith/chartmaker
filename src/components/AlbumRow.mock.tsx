const mock: typeof import('@/components/AlbumRow') = {
    AlbumRow: ({ children }) =>
        <div className='mock-album-row'>
            {children}
        </div>
}


export const AlbumRow = mock.AlbumRow
