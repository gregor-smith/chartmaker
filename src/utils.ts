import html2canvas from 'html2canvas'

import { BACKGROUND_COLOUR } from './style'


export function readFileText(file: File): Promise<string> {
    return new Promise((resolve, reject) => {
        const reader = new FileReader()
        reader.addEventListener('error', reject)
        reader.addEventListener('load', () => resolve(reader.result as string))
        reader.readAsText(file, 'utf-8')
    })
}


export function findIndex<T>(array: ReadonlyArray<T>, predicate: (item: T) => boolean): number | null {
    const index = array.findIndex(predicate)
    return index === -1 ? null : index
}


export async function elementToDataURI(element: HTMLElement, scale: number) {
    const canvas = await html2canvas(element, {
        allowTaint: true,
        scale,
        useCORS: true,
        imageTimeout: 0,
        backgroundColor: BACKGROUND_COLOUR
    })
    return canvas.toDataURL()
}


export function jsonToDataURI(json: string): string {
    return 'data:application/json;charset=utf-8,' + encodeURIComponent(json)
}


export function downloadURI(uri: string, filename: string) {
    const link = document.createElement('a')
    link.href = uri
    link.download = filename
    link.click()
    link.remove()
}
