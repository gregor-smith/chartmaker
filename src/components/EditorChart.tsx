import type { FC, Ref } from 'react'

import type { Chart as ChartDetails, Album, NamedAlbum } from '@/types'
import type { DispatchProps } from '@/reducer'
import { EditorAlbumRow } from '@/components/EditorAlbumRow'
import { EditorAlbumTitleGroup } from '@/components/EditorAlbumTitleGroup'
import { Chart, AlbumRowComponentProps, TitleGroupComponentProps } from '@/components/Chart'
import { identifiedAlbumIsPlaceholder } from '@/state'


function isNamedAlbum(album: Album): album is NamedAlbum {
    return !identifiedAlbumIsPlaceholder(album)
}


export type EditorChartProps =
    & DispatchProps
    & ChartDetails
    & {
        innerRef: Ref<HTMLElement>
        highlighted: number | undefined
    }


export const EditorChart: FC<EditorChartProps> = ({ dispatch, highlighted, ...props }) => {
    const AlbumRowComponent: FC<AlbumRowComponentProps<Album>> = props =>
        <EditorAlbumRow {...props}
            dispatch={dispatch}
            highlighted={highlighted}/>
    const TitleGroupComponent: FC<TitleGroupComponentProps<NamedAlbum>> = props =>
        <EditorAlbumTitleGroup {...props}
            dispatch={dispatch}
            highlighted={highlighted}/>

    return (
        <Chart {...props}
            dispatch={dispatch}
            albumRowComponent={AlbumRowComponent}
            titleGroupComponent={TitleGroupComponent}
            isNamedAlbum={isNamedAlbum}/>
    )
}
