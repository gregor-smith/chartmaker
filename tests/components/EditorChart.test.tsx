import { render } from 'react-dom'

import type { ChartShape } from '@/types'
import { EditorChart } from '@/components/EditorChart'

import { RenderContainer, ignore, createTestNamedAlbums } from '../utils'


jest.mock('@/components/EditorAlbumRow')
jest.mock('@/components/EditorAlbumTitleGroup')
jest.mock('@/components/Chart')


const container = new RenderContainer()


test.each<[ string, ChartShape, number, number, number | undefined ]>([
    [ 'Test chart', { tag: 'Top', size: 100 }, 1, 2, 1 ],
    [ 'Test chart 2', { tag: 'Collage' }, 3, 4, undefined ],
])('renders chart with editor rows and titles', (name, shape, rowsX, rowsY, highlighted) => {
    render(
        <EditorChart albums={createTestNamedAlbums(5)}
            name={name}
            shape={shape}
            rowsX={rowsX}
            rowsY={rowsY}
            innerRef={ignore}
            dispatch={ignore}
            highlighted={highlighted}/>,
        container.element
    )

    expect(container.element).toMatchSnapshot()
})
