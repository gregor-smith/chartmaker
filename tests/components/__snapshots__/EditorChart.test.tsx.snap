// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders chart with editor rows and titles 1`] = `
<div
  id="test-render-container"
>
  <div
    class="mock-chart"
  >
    Albums: [
  {
    "id": 1,
    "name": "Test album 1",
    "url": "https://test.com/1"
  },
  {
    "id": 2,
    "name": "Test album 2",
    "url": "https://test.com/2"
  },
  {
    "id": 3,
    "name": "Test album 3",
    "url": "https://test.com/3"
  },
  {
    "id": 4,
    "name": "Test album 4",
    "url": "https://test.com/4"
  },
  {
    "id": 5,
    "name": "Test album 5",
    "url": "https://test.com/5"
  }
]
    Name: Test chart
    Rows X: 1
    Rows Y: 2
    Shape: {"tag":"Top","size":100}
    <div
      class="mock-editor-album-title"
    >
      Albums: []
      Size: Mock chart size
      Highlighted: 1
    </div>
    <div
      class="mock-editor-album-title-group"
    >
      Albums: []
      Highlighted: 1
    </div>
  </div>
</div>
`;

exports[`renders chart with editor rows and titles 2`] = `
<div
  id="test-render-container"
>
  <div
    class="mock-chart"
  >
    Albums: [
  {
    "id": 1,
    "name": "Test album 1",
    "url": "https://test.com/1"
  },
  {
    "id": 2,
    "name": "Test album 2",
    "url": "https://test.com/2"
  },
  {
    "id": 3,
    "name": "Test album 3",
    "url": "https://test.com/3"
  },
  {
    "id": 4,
    "name": "Test album 4",
    "url": "https://test.com/4"
  },
  {
    "id": 5,
    "name": "Test album 5",
    "url": "https://test.com/5"
  }
]
    Name: Test chart 2
    Rows X: 3
    Rows Y: 4
    Shape: {"tag":"Collage"}
    <div
      class="mock-editor-album-title"
    >
      Albums: []
      Size: Mock chart size
      Highlighted: undefined
    </div>
    <div
      class="mock-editor-album-title-group"
    >
      Albums: []
      Highlighted: undefined
    </div>
  </div>
</div>
`;
